samza-api/src/main/java/org/apache/samza/container/grouper/stream/StreamPartitionMapper.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.container.grouper.stream;
+
+import org.apache.samza.system.SystemStreamPartition;
+
+/**
+ * Input streams of a samza job can be either expanded or contracted by the user.
+ * This abstraction determines the previous {@link SystemStreamPartition} for a {@link SystemStreamPartition}
+ * of a input stream after the stream expansion. This will be used in {@link SystemStreamPartitionGrouper}
+ * implementations to generate partition expansion aware task assignments.
+ */
+public interface StreamPartitionMapper {
```

Comment:
Why a new interface instead of a new config/mode for the existing grouper? What other implementations of this interface do you envision?

Reply:
When partition expansion or contraction happens on a input stream, we want to know the producer partitioning function used in the upstream to map the partitions back to correct task.  


1. This interface is for abstracting out that producer partitioning function part in samza for different input sources. `DefaultPartitioner`, `MurmurPartitioner` and `RangePartitioner` are some of the kafka producer partitioning functions  used within  LI through samza, tracker and other libraries. 

2. In SEP-5, this partition mapping was made pluggable as well.